_posts/2018-06-18-configuration.md

```diff
@@ -0,0 +1,160 @@
+---
+author: Sean Callan
+author_link: https://github.com/doomspork
+categories: general
+date: 2018-06-18
+layout: post
+title:  Confusing Configuration
+excerpt: >
+  We attempt to clear up some confusion around configuration by looking at the different types, the roles they play, and a different approach we could take.
+---
+
+There's been a lot of discussion about configuration in the community lately.
+We thought this would be an opportune time to discuss configuration and how best to handle it within an Elixir application.
+It is surprising to see how a small change to our applications configuration can eliminate much of the headaches others are experiencing._
+
+### Configuration types
+
+Before we go any further, let's look at the two configuration types and the roles they play.
+
+__Runtime Configuration__
+
+If you've ever used a system environment variable to configure some part of your application, then you're familiar with runtime configuration.
+As the name suggests, this is the configuration for an application at the time it is run.
+These are the values we can expect to change as we deploy our build artifacts to different systems.
+
+__Application Configuration__
+
+Application configuration is something different, and the difference, while subtle, can be a pitfall in certain situations.
+The difference shines when we consider that our code, and it's configuration, is compiled into a build artifact we can distribute.
+
+We can say with certainty that no matter where our application is run there are certain things we want to remain constant.
+No matter where we deploy our application, we want know we want to use the same `Logger` configuration.
+If we're relying on dependency injection for testing, then we know for certain we don't want to use those dependencies in the final deliverable.
+They configure the function of our code.
+
+### How it's been done
+
+The frustration for many people has to do with the usage of `Application.get_env/2` and `System.get_env/1`.
+We'll jump right in by looking at a configuration that's common to many Elixir projects:
+
+```elixir
+use Mix.Config
+
+config :example_app, Data.Repo,
+  adapter: Ecto.Adapters.MySQL,
+  username: System.get_env("EXAMPLE_APP_USERNAME"),
+  password: System.get_env("EXAMPLE_APP_PASSWORD"),
+  hostname: System.get_env("EXAMPLE_APP_HOSTNAME"),
+  database: System.get_env("EXAMPLE_APP_DATABASE"),
+  pool_size: 10
+```
+
+Simple and harmless enough, right?
+Wrong!
+
+Our application's configuration, defined in `config.exs` and friends, is compiled when we generate build artifacts, like those produced by Distillery.
+That means those `System.get_env/1` functions need to be resolved at compile time.
+See the problem?
+Our application's code is coupled to our system configuration.
```

Comment:
Maybe it would be clearer to say that our application's _compiled_ code is coupled to the configuration of the system that compiled the code. I don't want to split hairs, but the use of `System.get_env/1` will always make the _behavior_ of the application dependent upon some system configuration, and I'd bet some people see behavior and code as synonymous.